### ISSUE TITLE
the kube-proxy cannot run on centos

### ISSUE THREAD

**Issue by anzhsoft** [kubernetes/kubernetes#23976]
when I run the cmd:
kube-proxy --logtostderr=true --v=0 --master=http://master:8080
I0407 17:22:24.978303   17582 server.go:216] Using userspace Proxier.
F0407 17:22:24.978441   17582 server.go:232] Unable to create proxier: failed to select a host interface: Unable to select an IP.

and i have checked the etcd and flanneld, it seems that they work as expected. 

very appreciated if anyone can help.

**Comment by zhouhaibing089** [comment]
I spent a little bit time to look at this, firstly, you should know where does the log come from:

1.  `Unable to create proxier`: https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/server.go#L232
2.  `failed to select a host interface`: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/userspace/proxier.go#L149
3.  `Unable to select an IP`: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/net/interface.go#L261

so following this code, the error happened when select an ip by reading the file `/proc/net/route`, I paste the code below:

```go
routes, err := getRoutes(inFile)
if err != nil {
	return nil, err
}
zero := net.IP{0, 0, 0, 0}
var finalIP net.IP
for i := range routes {
	//find interface with gateway
	if routes[i].Destination.Equal(zero) {
		glog.V(4).Infof("Default route transits interface %q", routes[i].Interface)
		finalIP, err := getIPFromInterface(routes[i].Interface, nw)
		if err != nil {
			return nil, err
		}
		if finalIP != nil {
			glog.V(4).Infof("Choosing IP %v ", finalIP)
			return finalIP, nil
		}
	}
}
glog.V(4).Infof("No valid IP found")
if finalIP == nil {
	return nil, fmt.Errorf("Unable to select an IP.")
}
return nil, nil
```

So basically you should check the file `/proc/net/route`, and see whether there is a route which has destination 0.0.0.0, if there is, then see whether this interface has an ip.

This is a sample content of that file:

```sh
$ cat /proc/net/route
Iface	Destination	Gateway 	Flags	RefCnt	Use	Metric	Mask		MTU	Window	IRTT
eth0	00000000	01F7410A	0003	0	0	0	00000000	0	0	0
tun0	0000070A	00000000	0001	0	0	0	0000FFFF	0	0	0
eth0	00F7410A	00000000	0001	0	0	0	00FFFFFF	0	0	0
eth0	FEA9FEA9	00000000	0005	0	0	0	FFFFFFFF	0	0	0
docker0	000011AC	00000000	0001	0	0	0	0000FFFF	0	0	0
```

another suggestion, you could use `--v=4` to see more logs.

**Comment by vishh** [comment]
cc @ArtfulCoder @kubernetes/sig-network

**Comment by anzhsoft** [comment]
thank you very much. After i added the route, the proxy can start.

**Comment by thockin** [comment]
So you did not have a default route?  I am curious how to handle this
better - if it happened to you it might happen to others...

**Comment by jsmucr** [comment]
It just did. :/

**Comment by mml** [comment]
@thockin a fresh occurence on #25050

**Comment by cmluciano** [comment]
Is this still an issue?
